src/hardware/gus.cpp

```diff
@@ -665,27 +673,37 @@ void Gus::ActivateVoices(uint8_t requested_voices)
 	}
 }
 
-AudioFrame Gus::RenderFrame()
+const std::vector<AudioFrame>& Gus::GetFrames(const int num_requested_frames)
 {
-	AudioFrame accumulator = {};
+	// use a static vector to avoid re-allocating each millisecond
+	static std::vector<AudioFrame> frames = {};
 
-	if (dac_enabled) {
+	// Size and zero the vector
+	frames.resize(check_cast<size_t>(num_requested_frames));
+	for (auto& frame : frames) {
+		frame = {0.0f, 0.0f};
+	}
 
+	if (dac_enabled) {
 		auto voice = voices.begin();
-		const auto voice_end = voice + active_voices;
-
-		while (voice < voice_end && *voice) {
-			const auto voice_frame = voice->get()->RenderFrame(
-			        ram, vol_scalars, pan_scalars);
-
-			accumulator.left += voice_frame.left;
-			accumulator.right += voice_frame.right;
-
+		const auto last_voice = voice + active_voices;
+		while (voice < last_voice && *voice) {
```

Comment:
Related: can't we move this into `Gus` and remove the `unique_ptr` cruft?

```cpp
using voice_array_t = std::array<std::unique_ptr<Voice>, MAX_VOICES>;
```

Reply:
willsnip the `unique_ptr` and moves to full objects.